Fix bad dSYM for symbol-stripped dynamic framework builds

friedbunny opened the pull request:
Fixes #6502. Stripped dynamic framework packages should now come with the correct dSYM.

Disabling `GCC_GENERATE_DEBUGGING_SYMBOLS` for `SYMBOLS=NO` builds meant that those builds had no debug symbols to strip or add to a dSYM. Running `make iframework SYMBOLS=NO` would fail to create a dSYM, but our deploy script would evidently [reuse the symbol-ful dSYM](https://github.com/mapbox/mapbox-gl-native/issues/6502#issuecomment-250605473).

With the added UUID validation, the above scenario would now fail:
```
* Stripping symbols from binaries…
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip: changes being made to the file will invalidate the code signature in: /Users/jason/Documents/Mapbox/mapbox-gl-native/build/ios/pkg/dynamic/Mapbox.framework/Mapbox (for architecture i386)
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip: changes being made to the file will invalidate the code signature in: /Users/jason/Documents/Mapbox/mapbox-gl-native/build/ios/pkg/dynamic/Mapbox.framework/Mapbox (for architecture x86_64)
* Validating dSYM and framework UUIDs
error: unable to open 'build/ios/pkg/dynamic/Mapbox.framework.dSYM/Contents/Resources/DWARF/Mapbox': No such file or directory
Error: dSYM and framework UUIDs do not match.
build/ios/pkg/dynamic/Mapbox.framework.dSYM/Contents/Resources/DWARF/Mapbox
  
build/ios/pkg/dynamic/Mapbox.framework/Mapbox
  23F302DF-17AB-384C-879B-746C8A73511A (arm64) 2877864D-E317-3250-8756-11D94C868177 (i386) 59352B56-89D5-3E8E-B83E-9B15657B5C83 (armv7) F91C2CAC-5283-3ADF-ACFB-DD86B4BF5BFF (x86_64)
make: *** [iframework] Error 1
```

Note: `GCC_GENERATE_DEBUGGING_SYMBOLS=YES` is the default setting.

### Useful resources
- https://developer.apple.com/library/content/technotes/tn2004/tn2123.html
- `nm Mapbox.framework/Mapbox | grep MGL` helped test symbol stripping.
- Archived [Xcode build settings reference](https://web.archive.org/web/20160424060222/https://developer.apple.com/library/ios/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW53).

/cc @boundsj @1ec5

friedbunny commented:
One other thing: the `strip` command was never being invoked because of [bash boolean problems](https://github.com/mapbox/mapbox-gl-native/pull/6531/files#diff-86a1df829aa488a1c4289ab852eb2961L182). Our symbol “stripping” was coming entirely from the fact that we weren’t generating debug symbols.

Now that we _are_ invoking `strip -Sx`, we’re also stripping non-global symbols. So far, this looks like it reduces the raw binary size of Mapbox.framework/Mapbox by 20% — v3.4.0-alpha.4 was 45.3MB, this PR is 35.1MB.

The dSYM size is also much smaller: 234MB vs 32MB. This worries me and I’m not sure why it’s happened, but I was able to successfully symbolicate crashes with this smaller dSYM...

1ec5 commented on platform/macos/scripts/package.sh:
```diff
@@ -43,7 +42,7 @@ if [[ -e ${PRODUCTS}/${BUILDTYPE}/${NAME}.framework.dSYM ]]; then
-if [[ "${GCC_GENERATE_DEBUGGING_SYMBOLS}" == false ]]; then
+if [[ ${SYMBOLS} = NO ]]; then
     step "Stripping binaries…"
     strip -Sx "${OUTPUT}/${NAME}.framework/${NAME}"
```
Would it be possible to do dSYM validation for macOS as well? I think it might’ve caught #5818 indirectly.

friedbunny replied:
Should be as easy as copying it over, I think. Will give it a try.

friedbunny replied:
Did this in 6723a611b02528b8474383c0f91e326e8839b4c3.

friedbunny commented:
- Added macOS dSYM validation.
- Added macOS packaging documentation.
- Went back and ensured that `make iframework BUILDTYPE=Debug` would skip dSYM validation, because there are no dSYMs generated (at least, not until #6529).

Should be ready to merge.

1ec5 commented:
Oh, this is probably worth mentioning in the iOS changelog too.
